presto-hudi/src/main/java/com/facebook/presto/hudi/ParquetPageSourceUtil.java

```diff
@@ -0,0 +1,210 @@
+import static com.facebook.presto.hudi.HudiErrorCode.HUDI_CANNOT_OPEN_SPLIT;
+import static com.facebook.presto.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static com.facebook.presto.parquet.ParquetTypeUtils.getColumnIO;
+import static com.facebook.presto.parquet.ParquetTypeUtils.getDescriptors;
+import static com.facebook.presto.parquet.ParquetTypeUtils.getParquetTypeByName;
+import static com.facebook.presto.parquet.predicate.PredicateUtils.buildPredicate;
+import static com.facebook.presto.parquet.predicate.PredicateUtils.predicateMatches;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static org.apache.parquet.io.ColumnIOConverter.constructField;
+
+final class ParquetPageSourceUtil
+{
+    private ParquetPageSourceUtil() {}
+
+    public static ConnectorPageSource createParquetPageSource(
+            TypeManager typeManager,
+            HdfsEnvironment hdfsEnvironment,
+            String user,
+            Configuration configuration,
+            Path path,
+            long start,
+            long length,
+            SchemaTableName tableName,
+            List<HudiColumnHandle> regularColumns,
+            boolean useParquetColumnNames,
+            boolean failOnCorruptedParquetStatistics,
+            DataSize maxReadBlockSize,
+            boolean batchReaderEnabled,
+            boolean verificationEnabled,
+            TupleDomain<HudiColumnHandle> effectivePredicate,
+            FileFormatDataSourceStats fileFormatDataSourceStats)
+    {
+        AggregatedMemoryContext systemMemoryContext = newSimpleAggregatedMemoryContext();
+
+        ParquetDataSource dataSource = null;
+        try {
+            ExtendedFileSystem filesystem = hdfsEnvironment.getFileSystem(user, path, configuration);
+            FileStatus fileStatus = filesystem.getFileStatus(path);
+            long fileSize = fileStatus.getLen();
+            long modificationTime = fileStatus.getModificationTime();
+            HiveFileContext hiveFileContext = new HiveFileContext(true, NO_CACHE_CONSTRAINTS,
+                    Optional.empty(), Optional.of(fileSize), modificationTime, false);
+            FSDataInputStream inputStream = filesystem.openFile(path, hiveFileContext);
+            dataSource = buildHdfsParquetDataSource(inputStream, path, fileFormatDataSourceStats);
+            ParquetMetadata parquetMetadata = MetadataReader.readFooter(dataSource, fileSize).getParquetMetadata();
+            FileMetaData fileMetaData = parquetMetadata.getFileMetaData();
+            MessageType fileSchema = fileMetaData.getSchema();
+
+            // Mapping from Hudi field ID to Parquet fields.
+            Map<Integer, Type> parquetIdToField = fileSchema.getFields().stream()
```

Comment:
Aren't we gonna use parquet column names? Why do we need to map based on ID? Also, i don't think hudi schema has ID field yet.

Reply:
Updated to a name based schema mapping.